synapse/lib/layer.py

```diff
@@ -4093,6 +4093,88 @@ async def iterNodeDataKeys(self, buid):
             prop = self.getAbrvProp(abrv)
             yield prop[0]
 
+    async def iterLayerAddPerms(self):
+
+        # nodes & props
+        for byts, abrv in self.propabrv.slab.scanByFull(db=self.propabrv.name2abrv):
+            form, prop = s_msgpack.un(byts)
+            if form is None:
+                continue
+
+            for _ in self.layrslab.scanByPref(abrv, db=self.byprop):
+                if prop:
+                    yield ('node', 'prop', 'set', f'{form}:{prop}')
+                else:
+                    yield ('node', 'add', form)
+
+                break
+
+        # tags
+        # NB: tag perms should be yielded for every leaf on every node in the layer
+        await self._saveDirtySodes()
+        for buid, byts in self.layrslab.scanByFull(db=self.bybuidv3):
```

Comment:
I think we should do this loop last since it will take the longest. I was hoping we could come up with some scheme to walk the tag indexes in parallel to detect root tags with no leaves but that might be more complicated than it is worth.